File: database-driver-api/src/main/java/de/lino/database/database/DatabaseSection.java
package de.lino.database.database;

import de.lino.database.database.entity.DatabaseEntry;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.UnmodifiableView;

import java.util.List;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;
import java.util.function.Consumer;

/**
 * Represents a single logical grouping of {@link DatabaseEntry} objects within a
 * {@link DatabaseProvider} (e.g. a SQL table, a MongoDB collection, a Redis key prefix or a
 * directory of JSON files) and exposes CRUD operations over its entries.
 * <p>
 * Every synchronous operation declared here has a corresponding {@code *Async} default method
 * that executes the same logic on the common {@link CompletableFuture} pool.
 */
public interface DatabaseSection {

    /**
     * Get the section's name.
     *
     * @return the section's name
     */
    String getName();

    /**
     * Insert a new json document into the database.
     *
     * @param databaseEntry the entry to insert
     */
    void insert(@NotNull DatabaseEntry databaseEntry);

    /**
     * Update an existing json document from the database.
     *
     * @param databaseEntry the entry to update
     */
    void update(@NotNull DatabaseEntry databaseEntry);

    /**
     * Delete an existing json document from the database.
     *
     * @param id primary key
     */
    void delete(@NotNull String id);

    /**
     * Count all existing json documents.
     *
     * @return the number of entries currently stored in this section
     */
    long count();

    /**
     * Clear this database section.
     */
    void clear();

    /**
     * Discards this section's own cached view of its entries and rebuilds it from the
     * backing store, picking up entries added, changed or removed by something other
     * than this section itself (e.g. a backup restored directly onto disk while this
     * section was already loaded).
     * <p>
     * Every implementation shipped by this module caches its entries in memory beyond
     * what each write already keeps in sync (see each implementation's own class-level
     * documentation), so a genuine re-read of the backing store is required here, not a
     * no-op - {@link #getEntries()} and friends would otherwise never reflect a change
     * made outside this section.
     */
    void reload();

    /**
     * Check whether a json document exists.
     *
     * @param id primary key
     * @return true, if json document can be found, otherwise false
     */
    boolean exists(@NotNull String id);

    /**
     * Find a matching json document from the database.
     *
     * @param id primary key
     * @return an {@link Optional} containing the matching {@link DatabaseEntry}, or empty if no
     * entry exists under the given id
     */
    Optional<DatabaseEntry> findEntryById(@NotNull String id);

    /**
     * Get an unmodifiable list of all database entities.
     * <p>
     * The whole section is materialized as one list, whatever its size - kept that way for
     * compatibility. A consumer working through a large section should prefer
     * {@link #forEachEntry(Consumer)} (constant memory) or {@link #getEntries(long, int)}
     * (one bounded page at a time) instead of holding every entry at once.
     *
     * @return an unmodifiable list of all entries currently stored in this section
     */
    @UnmodifiableView
    List<DatabaseEntry> getEntries();

    /**
     * Streams every entry of this section to {@code consumer}, one at a time, without
     * materializing the section as a whole - the constant-memory alternative to
     * {@link #getEntries()} for sections too large to hold in one list. No entry order is
     * guaranteed.
     * <p>
     * This {@code default} implementation exists only so {@link DatabaseSection}
     * implementations written before this method keep compiling - it simply iterates
     * {@link #getEntries()} and therefore still materializes everything. Every section shipped
     * by this library's plugin module overrides it with a genuinely streaming implementation.
     *
     * @param consumer called once per stored entry
     */
    default void forEachEntry(@NotNull Consumer<DatabaseEntry> consumer) {
        this.getEntries().forEach(consumer);
    }

    /**
     * A single page of this section's entries: the entries at positions
     * {@code [offset, offset + limit)} of a stable, id-ordered enumeration - the same page for
     * the same arguments as long as the data does not change, so a consumer can work through a
     * large section chunk by chunk without ever holding more than one page.
     * <p>
     * This {@code default} implementation exists only so {@link DatabaseSection}
     * implementations written before this method keep compiling - it slices
     * {@link #getEntries()} (in that list's order) and therefore still materializes
     * everything. Every section shipped by this library's plugin module overrides it with an
     * implementation that pushes the paging toward the backing store instead.
     *
     * @param offset how many entries of the enumeration to skip; must not be negative
     * @param limit  the most entries the page may hold; must not be negative
     * @return the page's entries, empty once {@code offset} lies beyond the section's end
     */
    default @UnmodifiableView List<DatabaseEntry> getEntries(long offset, int limit) {

        if (offset < 0) throw new IllegalArgumentException("@DatabaseSection.getEntries: offset must not be negative, got " + offset);
        if (limit < 0) throw new IllegalArgumentException("@DatabaseSection.getEntries: limit must not be negative, got " + limit);

        final List<DatabaseEntry> all = this.getEntries();
        if (limit == 0 || offset >= all.size()) return List.of();

        return List.copyOf(all.subList((int) offset, (int) Math.min(all.size(), offset + limit)));

    }

    /**
     * Execute the {@link #insert(DatabaseEntry)} process async.
     *
     * @param databaseEntry the entry to insert
     * @return a {@link CompletableFuture} that completes once the entry has been inserted
     */
    default CompletableFuture<Void> insertAsync(@NotNull DatabaseEntry databaseEntry) {
        return CompletableFuture.runAsync(() -> insert(databaseEntry));
    }

    /**
     * Execute the {@link #update(DatabaseEntry)} process async.
     *
     * @param databaseEntry the entry to update
     * @return a {@link CompletableFuture} that completes once the entry has been updated
     */
    default CompletableFuture<Void> updateAsync(@NotNull DatabaseEntry databaseEntry) {
        return CompletableFuture.runAsync(() -> update(databaseEntry));
    }

    /**
     * Execute the {@link #delete(String)} process async.
     *
     * @param id primary key
     * @return a {@link CompletableFuture} that completes once the entry has been deleted
     */
    default CompletableFuture<Void> deleteAsync(@NotNull String id) {
        return CompletableFuture.runAsync(() -> delete(id));
    }

    /**
     * Execute the {@link #count()} process async.
     *
     * @return a {@link CompletableFuture} resolving to the number of entries currently stored in
     * this section
     */
    default CompletableFuture<Long> countAsync() {
        return CompletableFuture.supplyAsync(this::count);
    }

    /**
     * Execute the {@link #clear()} section process async.
     *
     * @return a {@link CompletableFuture} that completes once the section has been cleared
     */
    default CompletableFuture<Void> clearAsync() {
        return CompletableFuture.runAsync(this::clear);
    }

    /**
     * Execute the {@link #exists(String)} process async.
     *
     * @param id primary key
     * @return a {@link CompletableFuture} resolving to {@code true} if the entry exists,
     * {@code false} otherwise
     */
    default CompletableFuture<Boolean> existsAsync(@NotNull String id) {
        return CompletableFuture.supplyAsync(() -> exists(id));
    }

    /**
     * Execute the {@link #findEntryById(String)} process async.
     *
     * @param id primary key
     * @return a {@link CompletableFuture} resolving to an {@link Optional} containing the
     * matching {@link DatabaseEntry}, or empty if none exists under the given id
     */
    default CompletableFuture<Optional<DatabaseEntry>> findEntryByIdAsync(@NotNull String id) {
        return CompletableFuture.supplyAsync(() -> findEntryById(id));
    }

    /**
     * Execute the {@link #getEntries()} process async.
     *
     * @return a {@link CompletableFuture} resolving to an unmodifiable list of all entries
     * currently stored in this section
     */
    default CompletableFuture<List<DatabaseEntry>> getEntriesAsync() {
        return CompletableFuture.supplyAsync(this::getEntries);
    }

    /**
     * Execute the {@link #forEachEntry(Consumer)} process async.
     *
     * @param consumer called once per stored entry, on the async pool's thread
     * @return a {@link CompletableFuture} that completes once every entry has been consumed
     */
    default CompletableFuture<Void> forEachEntryAsync(@NotNull Consumer<DatabaseEntry> consumer) {
        return CompletableFuture.runAsync(() -> forEachEntry(consumer));
    }

    /**
     * Execute the {@link #getEntries(long, int)} process async.
     *
     * @param offset how many entries of the enumeration to skip; must not be negative
     * @param limit  the most entries the page may hold; must not be negative
     * @return a {@link CompletableFuture} resolving to the page's entries
     */
    default CompletableFuture<List<DatabaseEntry>> getEntriesAsync(long offset, int limit) {
        return CompletableFuture.supplyAsync(() -> getEntries(offset, limit));
    }

    /**
     * Execute the {@link #reload()} process async.
     *
     * @return a {@link CompletableFuture} that completes once this section has been reloaded
     */
    default CompletableFuture<Void> reloadAsync() {
        return CompletableFuture.runAsync(this::reload);
    }

}

File: python/database-driver-api/src/database_driver/api/database/database_section.py
"""Mirror of ``de.lino.database.database.DatabaseSection``."""

from __future__ import annotations

import asyncio
from abc import ABC, abstractmethod
from collections.abc import Callable

from database_driver.api.database.entity.database_entry import DatabaseEntry


class DatabaseSection(ABC):
    """Represents a single logical grouping of ``DatabaseEntry`` objects within a
    ``DatabaseProvider`` (e.g. a SQL table, a MongoDB collection, a Redis key prefix or a
    directory of JSON files) and exposes CRUD operations over its entries.

    Every synchronous operation declared here has a corresponding ``*_async`` coroutine
    method that executes the same logic on a worker thread via ``asyncio.to_thread`` -
    the direct analogue of the Java edition's ``CompletableFuture.runAsync`` default
    methods, which likewise offload the blocking call to a shared pool rather than being
    natively non-blocking.
    """

    @abstractmethod
    def get_name(self) -> str:
        """Returns the section's name."""

    @abstractmethod
    def insert(self, database_entry: DatabaseEntry) -> None:
        """Inserts a new json document into the database.

        Raises:
            DataAlreadyExist: If an entry with the same id already exists.
        """

    @abstractmethod
    def update(self, database_entry: DatabaseEntry) -> None:
        """Updates an existing json document in the database.

        Raises:
            NoSuchEntryFound: If no entry exists under the given entry's id.
        """

    @abstractmethod
    def delete(self, id: str) -> None:
        """Deletes an existing json document from the database by primary key.

        Raises:
            NoSuchEntryFound: If no entry exists under ``id``.
        """

    @abstractmethod
    def count(self) -> int:
        """Returns the number of entries currently stored in this section."""

    @abstractmethod
    def clear(self) -> None:
        """Clears this database section."""

    @abstractmethod
    def reload(self) -> None:
        """Discards this section's own cached view of its entries and rebuilds it from
        the backing store, picking up entries added, changed or removed by something other
        than this section itself (e.g. a backup restored directly onto disk while this
        section was already loaded).

        Every implementation shipped by the plugin module caches its entries in memory
        beyond what each write already keeps in sync (see each implementation's own
        class-level documentation), so a genuine re-read of the backing store is required
        here, not a no-op - :meth:`get_entries` and friends would otherwise never reflect
        a change made outside this section.
        """

    @abstractmethod
    def exists(self, id: str) -> bool:
        """Returns whether a json document exists under the given primary key."""

    @abstractmethod
    def find_entry_by_id(self, id: str) -> DatabaseEntry | None:
        """Finds a matching json document by primary key.

        Returns:
            The matching ``DatabaseEntry``, or ``None`` if no entry exists under ``id`` -
            standing in for the Java edition's ``Optional``.
        """

    @abstractmethod
    def get_entries(self) -> list[DatabaseEntry]:
        """Returns a list of all database entries.

        The whole section is materialized as one list, whatever its size - kept that way
        for compatibility. A consumer working through a large section should prefer
        :meth:`for_each_entry` (constant memory) or :meth:`get_entries_page` (one bounded
        page at a time) instead of holding every entry at once.
        """

    def for_each_entry(self, consumer: Callable[[DatabaseEntry], None]) -> None:
        """Streams every entry of this section to ``consumer``, one at a time, without
        materializing the section as a whole - the constant-memory alternative to
        :meth:`get_entries` for sections too large to hold in one list. No entry order is
        guaranteed.

        This default implementation exists only so ``DatabaseSection`` implementations
        written before this method keep working - it simply iterates
        :meth:`get_entries` and therefore still materializes everything. Every section
        shipped by the plugin module overrides it with a genuinely streaming
        implementation.
        """
        for entry in self.get_entries():
            consumer(entry)

    def get_entries_page(self, offset: int, limit: int) -> list[DatabaseEntry]:
        """A single page of this section's entries: the entries at positions
        ``[offset, offset + limit)`` of a stable, id-ordered enumeration - the same page
        for the same arguments as long as the data does not change, so a consumer can
        work through a large section chunk by chunk without ever holding more than one
        page.

        This default implementation exists only so ``DatabaseSection`` implementations
        written before this method keep working - it slices :meth:`get_entries` (in that
        list's order) and therefore still materializes everything. Every section shipped
        by the plugin module overrides it with an implementation that pushes the paging
        toward the backing store instead.

        Raises:
            ValueError: If ``offset`` or ``limit`` is negative.
        """
        if offset < 0:
            raise ValueError(f"@DatabaseSection.get_entries_page: offset must not be negative, got {offset}")
        if limit < 0:
            raise ValueError(f"@DatabaseSection.get_entries_page: limit must not be negative, got {limit}")

        all_entries = self.get_entries()
        if limit == 0 or offset >= len(all_entries):
            return []
        return list(all_entries[offset : offset + limit])

    # ------------------------------------------------------------------ async

    async def insert_async(self, database_entry: DatabaseEntry) -> None:
        """Executes the :meth:`insert` process async."""
        await asyncio.to_thread(self.insert, database_entry)

    async def update_async(self, database_entry: DatabaseEntry) -> None:
        """Executes the :meth:`update` process async."""
        await asyncio.to_thread(self.update, database_entry)

    async def delete_async(self, id: str) -> None:
        """Executes the :meth:`delete` process async."""
        await asyncio.to_thread(self.delete, id)

    async def count_async(self) -> int:
        """Executes the :meth:`count` process async."""
        return await asyncio.to_thread(self.count)

    async def clear_async(self) -> None:
        """Executes the :meth:`clear` section process async."""
        await asyncio.to_thread(self.clear)

    async def exists_async(self, id: str) -> bool:
        """Executes the :meth:`exists` process async."""
        return await asyncio.to_thread(self.exists, id)

    async def find_entry_by_id_async(self, id: str) -> DatabaseEntry | None:
        """Executes the :meth:`find_entry_by_id` process async."""
        return await asyncio.to_thread(self.find_entry_by_id, id)

    async def get_entries_async(self) -> list[DatabaseEntry]:
        """Executes the :meth:`get_entries` process async."""
        return await asyncio.to_thread(self.get_entries)

    async def for_each_entry_async(self, consumer: Callable[[DatabaseEntry], None]) -> None:
        """Executes the :meth:`for_each_entry` process async; ``consumer`` runs on the
        worker thread."""
        await asyncio.to_thread(self.for_each_entry, consumer)

    async def get_entries_page_async(self, offset: int, limit: int) -> list[DatabaseEntry]:
        """Executes the :meth:`get_entries_page` process async."""
        return await asyncio.to_thread(self.get_entries_page, offset, limit)

    async def reload_async(self) -> None:
        """Executes the :meth:`reload` process async."""
        await asyncio.to_thread(self.reload)
